extensions/amp-link-rewriter/0.1/link-shifter.js

```diff
@@ -0,0 +1,227 @@
+
+    this.viewer_.getReferrerUrl().then(referrerUrl => {
+      const pageAttributes = {
+        referrer: referrerUrl,
+        location: this.viewer_.getResolvedViewerUrl(),
+      };
+
+      htmlElement.href = this.replacePlaceHolders(htmlElement, pageAttributes);
+
+      // If the link has been "activated" via contextmenu,
+      // we have to keep the shifting in mind
+      if (this.event_.type === 'contextmenu') {
+        this.ctxAttrValue_ = CTX_ATTR_VALUE().toString();
+        htmlElement.setAttribute(CTX_ATTR_NAME, this.ctxAttrValue_);
+      }
+
+      this.viewer_.win.setTimeout(() => {
+        htmlElement.href = oldValHref;
+
+        if (htmlElement.hasAttribute(CTX_ATTR_NAME)) {
+          htmlElement.removeAttribute(CTX_ATTR_NAME);
+        }
+
+      }, ((this.event_.type === 'contextmenu') ? 15000 : 500));
+    });
+  }
+
+  /**
+   * @param {?Element} htmlElement
+   * @param {!Object} pageAttributes
+   * @return {string}
+   */
+  replacePlaceHolders(htmlElement, pageAttributes) {
+    const {vars} = this.configOpts_;
+    let {output} = this.configOpts_;
+    const data = {};
+
+    /**
+     * Replace placeholders for anchor attributes
+     * defined in white list constant array
+     */
+    WL_ANCHOR_ATTR.forEach(val => {
+      let attrValue = '';
+
+      if (htmlElement.getAttribute(val)) {
+        attrValue = htmlElement.getAttribute(val);
+      }
+
+      output = output.replace('${' + val + '}', encodeURIComponent(attrValue));
+    });
+
+    /**
+     * Replace placeholders for properties of the document
+     * at the moment only referrer and location
+     */
+    Object.keys(pageAttributes).forEach(key => {
+      let propValue = '';
+
+      if (pageAttributes[key]) {
+        propValue = pageAttributes[key];
+      }
+
+      output = output.replace('${' + key + '}', encodeURIComponent(propValue));
+    });
+
+    /**
+     * Set on data object properties and values defined
+     * on 'vars config property'
+     */
+    Object.keys(vars).forEach(key => {
+      let confValue = '';
+
+      if (vars[key]) {
+        confValue = vars[key];
+      }
+
+      data[key] = confValue;
+    });
+
+    /**
+     * Set on data object properties and values set on the element
+     * 'data attributes' in case these have the save name that the
+     * 'vars config property', 'data attributes' values will
+     * overwrite 'vars config values'
+     */
+    const {dataset} = htmlElement;
+    Object.keys(dataset).forEach(key => {
+      const dataValue = dataset[key];
+      let dataProp = key.split(PREFIX_DATA_ATTR)[1];
+
+      dataProp = dataProp.replace(/^[A-Z]/, match => {
+        return match.toLowerCase();
+      });
+
+      data[dataProp] = dataValue;
+    });
+
+    /**
+     * Replace placeholders under data namespace for values merged of
+     * 'vars config property' and 'data attributes'
+     */
+
+    Object.keys(data).forEach(key => {
+      const placeHoder = '${' + NS_DATA_PH + key + '}';
```

Comment:
to be consistent with `amp-analytics`, let's we access data vars using
`${fooBar}` instead of `${data.fooBar}`